phylogeny/prototypeSelection.py

```
@@ -0,0 +1,99 @@
+    '''Compute the sum of pairwise distances for the given elements according to
+    the given distance matrix.
+
+    Parameters
+    ----------
+    elements: sequence of str
+        list or elements for which the sum of distances is computed
+    dm: skbio.stats.distance.DistanceMatrix
+        pairwise distance matrix.
+
+    Returns
+    -------
+    float:
+        the sum of all pairwise distances of dm for IDs in elements
+    '''
+
+    return np.tril(dm.filter(elements).data).sum()
+
+
+def prototypeSelection_exhaustive(
+      distanceMatrix: DistanceMatrix,
+      numPrototypes: int, maxCombinationsToTest: int=200000) -> Sequence[str]:
+    '''Select k prototypes for given distance matrix
+
+    Parameters
+    ----------
+    distanceMatrix: skbio.stats.distance.DistanceMatrix
+        pairwise distances for all elements in the full set S.
+        Must be symmetric and non-hollow.
+    numPrototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+    maxCombinationsToTest: int
+        The maximal number of combinations to test. If exceeding, the function
+        declines execution.
+
+    Returns
+    -------
+    sequence of str
+        A sequence holding selected prototypes, i.e. a sub-set of the
+        elements in the distance matrix.
+
+    Notes
+    -----
+    This is the reference implementation for an exact algorithm for the
+    prototype selection problem. It has an exponential runtime and will only
+    operate on small instances (< maxCombinationsToTest).
+    Idea: test all (n over k) combinations of selecting k elements from n with-
+          out replacement. Compute the objective for each such combination and
+          report the combination with maximal value.
+    '''
+    assert(numPrototypes >= 2)
```

How often do you expect this asserts to be failing? Ideally you want to raise a specific exception with a meaningful messages, so other developers understand what is going on with the code (applies to all)

---

True, but I do not really know what the best matching Exception in this case would be.